File: leetcode/LC/_c++code/1448_good_nodes.cpp
#include <climits>
#include <iostream>
using namespace std;
#include <vector>

struct TreeNode {
      int val;
     TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/* 

LC 1448
pass max possible by value and store result in reference..
*/
class Solution {
public:
    void goodNodesHelper(TreeNode* root, int &res, int max){
        if(root == NULL)
            return;
        if(root->val >= max){
            res++;
            max = root->val;
        }
        goodNodesHelper(root->left, res, max);
        goodNodesHelper(root->right, res, max);
        
    }
    int goodNodes(TreeNode* root) {
        int res = 0;
        goodNodesHelper(root, res, INT_MIN);
        return res;
    }
};

File: leetcode/LC/tree/1448_good_nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodeHelper(self, root, maxallowed, res):
        if not root:
            return
        if root.val >= maxallowed:
            res[0] += 1
            maxallowed = root.val
        self.goodNodeHelper(root.left, maxallowed, res)
        self.goodNodeHelper(root.right, maxallowed, res)
        
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        res = [0]
        self.goodNodeHelper(root, float('-inf'), res)
        return res[0]
